internal/statestore/redis.go

```diff
@@ -476,78 +476,83 @@ func (rb *redisBackend) GetTickets(ctx context.Context, ids []string) ([]*pb.Tic
 // This function guarantees if any of the input ids does not exists, the state of the storage service won't be altered.
 // However, since Redis does not support transaction roll backs (see https://redis.io/topics/transactions), some of the
 // assignment fields might be partially updated if this function encounters an error halfway through the execution.
-func (rb *redisBackend) UpdateAssignments(ctx context.Context, ids []string, assignment *pb.Assignment) error {
-	if assignment == nil {
-		return status.Error(codes.InvalidArgument, "assignment is nil")
-	}
-	if len(ids) == 0 {
-		return nil
+func (rb *redisBackend) UpdateAssignments(ctx context.Context, req *pb.AssignTicketsRequest) (*pb.AssignTicketsResponse, error) {
+	resp := &pb.AssignTicketsResponse{}
+	if len(req.Assignments) == 0 {
+		return resp, nil
+	}
+
+	idToA := make(map[string]*pb.Assignment)
+	ids := make([]string, 0)
+	idsI := make([]interface{}, 0)
+	for _, a := range req.Assignments {
+		if a.Assignment == nil {
+			return nil, status.Error(codes.InvalidArgument, "AssignmentGroup.Assignment is required")
+		}
+
+		for _, id := range a.TicketIds {
+			if _, ok := idToA[id]; ok {
+				return nil, status.Errorf(codes.InvalidArgument, "Ticket id %s is assigned multiple times in one assign tickets call.", id)
+			}
+
+			idToA[id] = a.Assignment
+			ids = append(ids, id)
+			idsI = append(idsI, id)
+		}
 	}
 
 	redisConn, err := rb.connect(ctx)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer handleConnectionClose(&redisConn)
 
-	b := make([]interface{}, len(ids))
-	for i := range ids {
-		b[i] = ids[i]
-	}
-
-	ticketBytes, err := redis.ByteSlices(redisConn.Do("MGET", b...))
+	ticketBytes, err := redis.ByteSlices(redisConn.Do("MGET", idsI...))
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	tickets := make([]*pb.Ticket, 0, len(ticketBytes))
 	for i, ticketByte := range ticketBytes {
 		// Tickets may be deleted by the time we read it from redis.
 		if ticketByte == nil {
-			return status.Errorf(codes.NotFound, "failed to get ticket %s from statestore", ids[i])
-		}
-		t := &pb.Ticket{}
-		err = proto.Unmarshal(ticketByte, t)
-		if err != nil {
-			redisLogger.WithFields(logrus.Fields{
-				"key": ids[i],
-			}).WithError(err).Error("Failed to unmarshal ticket from redis.")
-			return status.Errorf(codes.Internal, "%v", err)
+			resp.Failed = append(resp.Failed, &pb.AssignmentFailure{
+				TicketId: ids[i],
```

Comment:
`idsI[i]` works at here. do we really need the `ids` slice?

Reply:
It's also used to create an AssignmentFailure, which needs a string.

Reply:
is type assertion not working?

Reply:
I'd rather not use a type assertion if I don't have to.